src/packages/cli/src/cli.ts

```diff
@@ -56,45 +57,68 @@ process.on("uncaughtException", function (e) {
   }
 });
 
-// See http://stackoverflow.com/questions/10021373/what-is-the-windows-equivalent-of-process-onsigint-in-node-js
-if (process.platform === "win32") {
-  require("readline")
-    .createInterface({
-      input: process.stdin,
-      output: process.stdout
-    })
-    .on("SIGINT", function () {
-      process.emit("SIGINT" as any); // TODO: don't abuse process's emit
-    });
-}
-
-const closeHandler = async function () {
+let receivedShutdownSignal: boolean = false;
+const handleSignal = async (signal: NodeJS.Signals) => {
+  console.log(`Received shutdown signal: ${signal}`);
+  closeHandler();
+};
+const closeHandler = async () => {
   try {
     // graceful shutdown
-    if (server.status === 1) {
-      await server.close();
+    switch (server.status) {
+      case Status.opening:
+        receivedShutdownSignal = true;
+        console.log("Server is currently starting; waiting…");
+        return;
+      case Status.open:
+        console.log("Shutting down…");
+        await server.close();
+        console.log("Server has been shut down");
+        break;
```

Comment:
If we are currently opening it's best to wait until we're are fully opened before shutting down to avoid the possibility of DB corruption (or a db-induced SEGFAULT, which can happen!).

Thinking about it more, we actually do already guard against this here: https://github.com/trufflesuite/ganache-core/blob/8a7ba780a664e32b2210e32c3bfebe359e7cfc37/src/chains/ethereum/ethereum/src/blockchain.ts#L1128-L1134

So maybe this whole `Status.opening` case should be removed and we just let the internals take care of the problem. Yeah, I'm 95% sure external layers (like the CLI) shouldn't be concerned with safe-guarding the internals of the provider. I'd appreciate some other thoughts on this.

Reply:
I think either method is fine. I don't see an issue with extra safe guards. The logic you have also gives extra logging so the user knows what's happening. I prefer the way you have it actually